Refactor MultiCompilerHost

not-my-profile opened the pull request.

andrewbranch commented:
If you’d like to contribute, please comment on an issue or open a new one. Thanks!

not-my-profile commented:
Right for changes that aren't refactors or larger changes I usually open an issue first but for smaller refactors I don't really see the point, since the changes can just as well be discussed in the PR.

I was trying to make sense of the MultiCompilerHost implementation and found the code repetitive and hard to read, so I ended up doing this refactor (which I tried to explain in the last commit message).

andrewbranch commented:
Sorry to jump the gun a bit on closing this. When I saw a PR with zero description I assumed there was very little chance this would be valuable. I like this over all. I’m not a huge fan of classes, and the switch from closures to classes feels a bit arbitrary. I wouldn’t unconditionally reject classes, but `const that = this` and `.bind(this.traceCollector)` makes them feel not worthwhile in this case. If you’re willing to revert back to closures + POJOs, I’d be happy to take this reorganization.

not-my-profile commented:
Thanks for taking a look at the changes. I just rebased the PR. I agree about the two ugly points you spotted, but they are not due to the switch to classes. I just fixed both of them by using arrow functions (which you apparently can also do in a class).

not-my-profile commented on packages/core/src/multiCompilerHost.ts:
```diff
@@ -253,19 +177,40 @@ export function createMultiCompilerHost(fs: Package): MultiCompilerHost {
+  }
+}
+
+class TraceCollector {
+  private traces: string[] = [];
+
+  trace = (message: string) => {
+    this.traces.push(message);
+  };
```
I honestly didn't know that you could use an arrow function like this in a class to avoid having to `bind` the function outside but it works.

andrewbranch replied:
Methods are declared on the class prototype so they can be shared between all instances. This is a class field, which is initialized on each instance during construction (equivalent to doing `Object.defineProperty(this, "trace", ...)` in the constructor). This has a perf/memory cost as compared to a true method, but of course, the cost is equivalent to every other way of accomplishing what we’re trying to accomplish here—`.bind` also creates a new copy of the function; it was just happening at a different time and place. There’s no way around creating a new function identity for each `CompilerHost["trace"]`. (As a piece of trivia, this was a common trick for React devs in the days of class components for event handlers, so you could do `<button onClick={this.onClick}>` instead of `this.onClick.bind(this)`.) This still has a bit of a weird smell to me since you have to recognize which functions need this weird syntactic trick and which don’t. On the upside though, it’s a bit more efficient since at least `read` and `clear` are shared between all instances. It’s mostly a personal preference. I don’t love it but I appreciate the contribution so I’ll merge it 😛
